backends/graphics/atari/atari-graphics.cpp

```diff
@@ -0,0 +1,998 @@
+	}
+
+	Mfree(_overlayScreen);
+	_overlayScreen = nullptr;
+}
+
+bool AtariGraphicsManager::hasFeature(OSystem::Feature f) const {
+	switch (f) {
+	case OSystem::Feature::kFeatureAspectRatioCorrection:
+		debug("hasFeature(kFeatureAspectRatioCorrection): %d", !_vgaMonitor);
+		return !_vgaMonitor;
+	case OSystem::Feature::kFeatureCursorPalette:
+		debug("hasFeature(kFeatureCursorPalette): %d", isOverlayVisible());
+		return true;
+	case OSystem::Feature::kFeatureVSync:
+		debug("hasFeature(kFeatureVSync): %d", _vsync);
+		return true;
+	default:
+		return false;
+	}
+}
+
+void AtariGraphicsManager::setFeatureState(OSystem::Feature f, bool enable) {
+	switch (f) {
+	case OSystem::Feature::kFeatureAspectRatioCorrection:
+		debug("setFeatureState(kFeatureAspectRatioCorrection): %d", enable);
+		_oldAspectRatioCorrection = _aspectRatioCorrection;
+		_aspectRatioCorrection = enable;
+		break;
+	case OSystem::Feature::kFeatureVSync:
+		debug("setFeatureState(kFeatureVSync): %d", enable);
+		_vsync = enable;
+		break;
+	default:
+		[[fallthrough]];
+	}
+}
+
+bool AtariGraphicsManager::getFeatureState(OSystem::Feature f) const {
+	switch (f) {
+	case OSystem::Feature::kFeatureAspectRatioCorrection:
+		//debug("getFeatureState(kFeatureAspectRatioCorrection): %d", _aspectRatioCorrection);
+		return _aspectRatioCorrection;
+	case OSystem::Feature::kFeatureCursorPalette:
+		//debug("getFeatureState(kFeatureCursorPalette): %d", isOverlayVisible());
+		return isOverlayVisible();
+	case OSystem::Feature::kFeatureVSync:
+		//debug("getFeatureState(kFeatureVSync): %d", _vsync);
+		return _vsync;
+	default:
+		return false;
+	}
+}
+
+bool AtariGraphicsManager::setGraphicsMode(int mode, uint flags) {
+	debug("setGraphicsMode: %d, %d", mode, flags);
+
+	if (mode >= 0 && mode <= 3) {
+		_pendingState.mode = (GraphicsMode)mode;
+		return true;
+	}
+
+	return false;
+}
+
+void AtariGraphicsManager::initSize(uint width, uint height, const Graphics::PixelFormat *format) {
+	debug("initSize: %d, %d, %d (vsync: %d, mode: %d)", width, height, format ? format->bytesPerPixel : 1, _vsync, (int)_pendingState.mode);
+
+	_pendingState.width = width;
+	_pendingState.height = height;
+	_pendingState.format = format ? *format : PIXELFORMAT8;
+}
+
+void AtariGraphicsManager::beginGFXTransaction() {
+	debug("beginGFXTransaction");
+}
+
+OSystem::TransactionError AtariGraphicsManager::endGFXTransaction() {
+	debug("endGFXTransaction");
+
+	int error = OSystem::TransactionError::kTransactionSuccess;
+
+	// always initialize (clear screen, mouse init, ...)
+	//if (_pendingState == _currentState)
+	//	return static_cast<OSystem::TransactionError>(error);
+
+	if (_pendingState.format != PIXELFORMAT8)
+		error |= OSystem::TransactionError::kTransactionFormatNotSupported;
+
+	if ((_pendingState.width != 320 || (_pendingState.height != 200 && _pendingState.height != 240))
+			&& (_pendingState.width != 640 || (_pendingState.height != 400 && _pendingState.height != 480)))
```

Comment:
Several engines support unusual resolutions like 256x240 (NES Maniac Mansion), 512x342 (MacVenture, WAGE) or 544x332 (Myst), which could potentially be supported by centring the game screen within a larger screen mode.

Reply:
That is a good point, thanks. As I plan some other changes in this direction in the next iteration, I put a TODO in there for now. You are right, it's not hard to change but right now the code is quite well tested so I don't want to break it.